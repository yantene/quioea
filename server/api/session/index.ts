import { UserInfoFull } from "$/types"
import { SessionCreateBody } from "$/validators"

export type Methods = {
  get: {
    reqHeaders: { cookie: string }
    resBody: UserInfoFull
  }
  put: {
    reqBody: SessionCreateBody
    status: 201
    resBody: UserInfoFull
    resHeaders: { "Set-Cookie": string }
  }
  delete: {
    status: 204
  }
}
