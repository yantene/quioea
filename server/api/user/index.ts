import type { AuthHeader, UserInfo } from "$/types"
import type { ReadStream } from "fs"

export type Methods = {
  get: {
    reqHeaders: AuthHeader
    resBody: UserInfo
  }

  post: {
    reqHeaders: AuthHeader
    reqFormat: FormData
    reqBody: { icon: File | ReadStream }
    resBody: UserInfo
  }
}
