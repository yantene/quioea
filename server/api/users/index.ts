import { UserInfo } from "$/types"
import { UserIndexQuery, UserCreateBody } from "$/validators"

export type Methods = {
  get: {
    query?: UserIndexQuery
    resBody: UserInfo[]
  }
  post: {
    reqBody: UserCreateBody
    status: 201
    resBody: UserInfo
  }
}
