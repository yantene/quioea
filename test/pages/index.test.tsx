import React from "react"
import { cache } from "swr"
import dotenv from "dotenv"
import Fastify, { FastifyInstance } from "fastify"
import cors from "fastify-cors"
import aspida from "@aspida/axios"
import api from "~/server/api/$api"
import Home from "~/pages/index"
import { render, fireEvent } from "../testUtils"

dotenv.config({ path: "server/.env" })

const apiClient = api(aspida(undefined, { baseURL: process.env.API_BASE_PATH }))
const res = function <T extends () => unknown>(
  data: ReturnType<T> extends Promise<infer S> ? S : never,
) {
  return data
}

let fastify: FastifyInstance

beforeAll(() => {
  fastify = Fastify()
  fastify.register(cors)
  fastify.get(apiClient.tasks.$path(), (_, reply) => {
    reply.send(
      res<typeof apiClient.tasks.$get>([
        { id: 1, label: "foo task", done: false },
        { id: 2, label: "bar task", done: true },
      ]),
    )
  })

  return fastify.listen(process.env.API_SERVER_PORT ?? 8080)
})

afterEach(() => cache.clear())
afterAll(() => fastify.close())

describe("Home page", () => {
  it("matches snapshot", async () => {
    const { asFragment, findByText } = render(<Home />, {})

    await findByText("foo task")
    expect(asFragment()).toMatchSnapshot()
  })

  it("clicking button triggers prompt", async () => {
    const { findByText } = render(<Home />, {})

    window.prompt = jest.fn()
    window.alert = jest.fn()

    fireEvent.click(await findByText("LOGIN"))
    expect(window.prompt).toHaveBeenCalledWith(
      "Enter the user id (See server/.env)",
    )
    expect(window.alert).toHaveBeenCalledWith("Login failed")
  })
})
