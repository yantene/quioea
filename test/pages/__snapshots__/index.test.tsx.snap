// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <main
      class="main"
    >
      <div
        class="userBanner"
      >
        <button>
          LOGIN
        </button>
      </div>
      <h1
        class="title"
      >
        Welcome to 
        <a
          href="https://nextjs.org"
        >
          Next.js!
        </a>
      </h1>
      <p
        class="description"
      >
        frourio-todo-app
      </p>
      <div>
        <form
          style="text-align: center;"
        >
          <input
            type="text"
            value=""
          />
          <input
            type="submit"
            value="ADD"
          />
        </form>
        <ul
          class="tasks"
        >
          <li>
            <label>
              <input
                type="checkbox"
              />
              <span>
                foo task
              </span>
            </label>
            <input
              style="float: right;"
              type="button"
              value="DELETE"
            />
          </li>
          <li>
            <label>
              <input
                checked=""
                type="checkbox"
              />
              <span>
                bar task
              </span>
            </label>
            <input
              style="float: right;"
              type="button"
              value="DELETE"
            />
          </li>
        </ul>
      </div>
    </main>
    <footer
      class="footer"
    >
      <a
        href="https://vercel.com?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
        rel="noopener noreferrer"
        target="_blank"
      >
        Powered by 
        <img
          alt="Vercel Logo"
          class="logo"
          src="/vercel.svg"
        />
      </a>
    </footer>
  </div>
</DocumentFragment>
`;
